server/routerlicious/packages/lambdas/src/muse/lambda.ts

```diff
@@ -0,0 +1,190 @@
+                if (!pendingMessages) {
+                    pendingMessages = [];
+                    this.pending.set(topic, pendingMessages);
+                }
+
+                pendingMessages.push(value);
+            }
+        }
+
+        this.pendingOffset = message;
+        this.sendPending();
+
+        return undefined;
+    }
+
+    public close() {
+        this.pending.clear();
+        this.current.clear();
+    }
+
+    private sendPending() {
+        // If there is work currently being sent or we have no pending work return early
+        if (this.current.size > 0 || this.pending.size === 0) {
+            return;
+        }
+
+        // Swap current and pending
+        const temp = this.current;
+        this.current = this.pending;
+        this.pending = temp;
+        const batchOffset = this.pendingOffset;
+
+        const allProcessed: Promise<void[]>[] = [];
+
+        // Process all the batches + checkpoint
+        for (const [, messages] of this.current) {
+            const processP = this.processMuseCoreParallel(messages);
+            allProcessed.push(processP);
+        }
+
+        Promise.all(allProcessed).then(
+            () => {
+                this.current.clear();
+                this.context.checkpoint(batchOffset as IQueuedMessage);
+                this.sendPending();
+            },
+            (error) => {
+                this.context.error(error, { restart: true });
+            });
+    }
+
+    private createDerivedGuid(referenceGuid: string, identifier: string) {
+        const hexHash = enc.Hex.stringify(SHA1(`${referenceGuid}:${identifier}`));
+        return `
+            ${hexHash.substr(0, 8)}-
+            ${hexHash.substr(8, 4)}-
+            ${hexHash.substr(12, 4)}-
+            ${hexHash.substr(16, 4)}-
+            ${hexHash.substr(20, 12)}`;
+    }
+
+    private async processMuseCoreParallel(messages: ISequencedOperationMessage[]) {
+        const processedMessages: Map<string, Promise<void>> = new Map();
+
+        for (const message of messages) {
+            if (message?.operation?.type === "op") {
+                const contents = JSON.parse(message.operation.contents);
+                const opData = contents.contents?.contents?.content?.contents;
+                if (opData && opData.op === 0 && opData.changeSet !== undefined) {
+                    // At this point is checked to be submitted to Muse
+                    const branchGuid: string = contents.contents.contents.content.address;
+
+                    const currentProcessing = processedMessages.get(branchGuid);
+                    if (currentProcessing) {
+                        processedMessages.set(
+                            branchGuid,
+                            currentProcessing.then(async () => this.processMuseCore(branchGuid, opData, message)),
+                        );
+                    } else {
+                        processedMessages.set(branchGuid, Promise.resolve());
```

Comment:
shouldn't Promise.resolve() be `this.processMuseCore(branchGuid, opData, message)`? Otherwise the first message in the sequence won't be processed.